### DISCUSSION TITLE
Message handling without a dispatcher

### DISCUSSION THREAD

**Discussion by abel1502** [aiogram/aiogram#1289] [General]
Hello! I'm curious to know whether aiogram has a public interface to process messages directly, rather than by deferring it to the dispatcher? There are use cases where it could allow for much cleaner implementations than the current model. For instance, currently obtaining user repies requires an explicit state machine an manual passing of internal state, whereas it could be respresented significantly more neatly with an async function:
```python3
async def handle(message):
    ...
    await message.answer("Please select your something")
    # Something I'm interested in, but haven't found
    response = await message.chat.get_message()
    something = process(message.text)
    await message.chat.answer(f"Your something is {something} now")
    ...
```

Essentially, I'm looking for a way to await the reception an incoming message directly, as opposed to trying to set up the next callback and somehow pass the state I'd like to persist to it. Is there such a thing? If not, I'd also appreciate a pointer towards a different telegram API library that might be structured in a closer way to this.

**Comment by JrooTJunior** [comment]
you can use bot instance as you wish without the Dispatcher but your message sounds like you need to use finite state machine https://docs.aiogram.dev/en/dev-3.x/dispatcher/finite_state_machine/index.html

**Reply by abel1502** [reply]
I'm aware of aiogram's state machines (and have mentioned it in my question), however I am dissatisfied with them for this use case. Asynchronous functions have been invented specifically to facilitate representing state machines with associated data. (Every await point is a state; the locals carried across it are the associated data). The compiler under the hood converts evey asynchronous function into a state machine object.

An implicit state machine defined by an async function would've represented my desired behaviour much better than an explicit one does. The only potentially negative aspect is that the state will be lost on bot restart, but this isn't an issue for my usecase.

If what you're saying is that there is no intermediate layer between `Bot` and `Dispatcher` (i.e. something that encapsulates possing/webhooks and exposes an `asyncio.Queue` or the like), I can implement it over the `Dispatcher` abstraction, but it would be a rather ugly solution, essentially boiling down to taking a step backwards.

**Reply by JrooTJunior** [reply]
Technically you can do it manually without using the Dispatcher.
It is possible with using asyncio events and queue but this solution totally unstable in due to when you stop the bot process on the middle of conversation it can't be recovered to the same point after start and also this way is very hard to be scalable.

But if you have statically defined handlers tree for each state and answer of the user, each user state with transitions is explicitly defined and the state is stored outside of the bot process (for example in simple key-value storage like Redis or in any other database) you can easilly make a fault-tolerant, stateless and scalable application out of the box (until you break it by your hands with global varialbles).
aiogram provides it in as simple a way as possible so that it can all be implemented.
